akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/ServerSentEventParser.scala

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2015 Heiko Seeberger
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package akka.http
+package scaladsl
+package unmarshalling
+
+import akka.stream.stage.{ GraphStage, GraphStageLogic, InHandler, OutHandler }
+import akka.stream.{ Attributes, FlowShape, Inlet, Outlet }
+import akka.http.scaladsl.model.ServerSentEvent
+
+private object ServerSentEventParser {
+
+  final object PosInt {
+    def unapply(s: String): Option[Int] =
+      try { Some(s.trim.toInt) } catch { case _: NumberFormatException ⇒ None }
+  }
+
+  final class Builder {
+
+    private var data = Vector.empty[String]
+
+    private var eventType = Option.empty[String]
+
+    private var id = Option.empty[String]
+
+    private var retry = Option.empty[Int]
+
+    private var _size = 0
+
+    def appendData(value: String): Unit = {
+      _size += 5 + value.length
+      data :+= value
+    }
+
+    def setType(value: String): Unit = {
+      val oldSize = eventType.fold(0)(6 + _.length)
+      _size += 6 + value.length - oldSize
+      eventType = Some(value)
+    }
+
+    def setId(value: String): Unit = {
+      val oldSize = id.fold(0)(3 + _.length)
+      _size += 3 + value.length - oldSize
+      id = Some(value)
+    }
+
+    def setRetry(value: Int, length: Int): Unit = {
+      val oldSize = retry.fold(0)(6 + _.toString.length)
+      _size += 6 + length - oldSize
+      retry = Some(value)
+    }
+
+    def hasData: Boolean =
+      data.nonEmpty
+
+    def size: Int =
+      _size
+
+    def build(): ServerSentEvent =
+      ServerSentEvent(data.mkString("\n"), eventType, id, retry)
+
+    def reset(): Unit = {
+      data = Vector.empty[String]
+      eventType = None
+      id = None
+      retry = None
+      _size = 0
+    }
+  }
+
+  private final val Data = "data"
+
+  private final val EventType = "event"
+
+  private final val Id = "id"
+
+  private final val Retry = "retry"
+
+  private val field = """([^:]+): ?(.*)""".r
```

Comment:
Interesting rule to only capitalize static constants. :) By that logic you would also have to use lower case for `object` names because they are only lazily instantiated on first access...